pandas/core/indexers.py

```diff
@@ -419,10 +417,10 @@ def check_array_indexer(array: AnyArrayLike, indexer: Any) -> Any:
 
     dtype = indexer.dtype
     if is_bool_dtype(dtype):
-        try:
+        if is_extension_array_dtype(dtype):
+            indexer = indexer.to_numpy(dtype=bool, na_value=False)
+        else:
             indexer = np.asarray(indexer, dtype=bool)
-        except ValueError:
```

Comment:
What error message do you get now when masking with an object-dtype ndarray containg bools and NaNs?

Reply:
```python
In [3]: s = pd.Series(range(3))                                                                                                                                                    

In [4]: s                                                                                                                                                                          
Out[4]: 
0    0
1    1
2    2
dtype: int64

In [5]: mask = np.array([True, False, np.nan], dtype=object)                                                                                                                       

In [6]: mask                                                                                                                                                                       
Out[6]: array([True, False, nan], dtype=object)

In [7]: s[mask]                                                                                                                                                                    
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-b9628bf46992> in <module>
----> 1 s[mask]

~/pandas/pandas/core/series.py in __getitem__(self, key)
    877                 key = list(key)
    878 
--> 879             if com.is_bool_indexer(key):
    880                 key = check_bool_indexer(self.index, key)
    881                 return self._get_values(key)

~/pandas/pandas/core/common.py in is_bool_indexer(key)
    128                 na_msg = "Cannot mask with non-boolean array containing NA / NaN values"
    129                 if isna(key).any():
--> 130                     raise ValueError(na_msg)
    131                 return False
    132             return True

ValueError: Cannot mask with non-boolean array containing NA / NaN values
```

Reply:
Good to know, thanks.